File: backend/internal/provider/extensions/kuwo.go
package extensions

import (
	"fmt"
	"net/url"
	"regexp"
	"strconv"
	"strings"

	"github.com/guohuiyuan/go-music-dl/internal/provider/model"
)

const (
	kuwoSearchBaseURL = "http://search.kuwo.cn/r.s"
	kuwoListBaseURL   = "http://nplserver.kuwo.cn/pl.svc"
	kuwoBrowseBaseURL = "http://wapi.kuwo.cn/api/pc/classify/playlist"
)

var (
	kuwoPlaylistLinkPattern = regexp.MustCompile(`(?i)/playlist_detail/(\d+)`)
	kuwoAlbumLinkPattern    = regexp.MustCompile(`(?i)/album_detail/(\d+)`)
)

type Kuwo struct {
	Cookie string
}

func NewKuwo(cookie string) *Kuwo {
	return &Kuwo{Cookie: strings.TrimSpace(cookie)}
}

func (client *Kuwo) SearchPlaylist(keyword string) ([]model.Playlist, error) {
	payload, err := client.searchCollection(keyword, "playlist")
	if err != nil {
		return nil, err
	}
	return kuwoPlaylists(payload["abslist"]), nil
}

func (client *Kuwo) SearchAlbum(keyword string) ([]model.Playlist, error) {
	payload, err := client.searchCollection(keyword, "album")
	if err != nil {
		return nil, err
	}
	return kuwoAlbums(payload["albumlist"]), nil
}

func (client *Kuwo) searchCollection(keyword, collectionType string) (map[string]interface{}, error) {
	query := url.Values{
		"client": {"kt"}, "all": {strings.TrimSpace(keyword)}, "pn": {"0"}, "rn": {"30"},
		"uid": {"0"}, "ver": {"kwplayer_ar_9.2.2.1"}, "vipver": {"1"},
		"show_copyright_off": {"1"}, "newver": {"1"}, "ft": {collectionType},
		"cluster": {"0"}, "strategy": {"2012"}, "encoding": {"utf8"},
		"rformat": {"json"}, "mobi": {"1"},
	}
	return getSingleQuotedJSON(kuwoSearchBaseURL+"?"+query.Encode(), client.Cookie, nil)
}

func (client *Kuwo) GetPlaylistSongs(id string) ([]model.Song, error) {
	_, songs, err := client.Playlist(id)
	return songs, err
}

func (client *Kuwo) GetAlbumSongs(id string) ([]model.Song, error) {
	_, songs, err := client.Album(id)
	return songs, err
}

func (client *Kuwo) Playlist(id string) (*model.Playlist, []model.Song, error) {
	id = strings.TrimSpace(id)
	if id == "" {
		return nil, nil, fmt.Errorf("kuwo playlist id is empty")
	}
	query := url.Values{
		"op": {"getlistinfo"}, "pid": {id}, "pn": {"0"}, "rn": {"1000"},
		"encode": {"utf8"}, "keyset": {"pl2012"}, "vipver": {"MUSIC_9.1.1.2_BCS2"}, "newver": {"1"},
	}
	payload, err := getJSON(kuwoListBaseURL+"?"+query.Encode(), client.Cookie, nil)
	if err != nil {
		return nil, nil, err
	}
	playlist := kuwoPlaylistDetail(payload, id)
	songs := kuwoSongs(payload["musiclist"], playlist)
	if len(songs) == 0 {
		return nil, nil, errorsForEmpty("kuwo playlist")
	}
	if playlist.TrackCount == 0 {
		playlist.TrackCount = len(songs)
	}
	return &playlist, songs, nil
}

func (client *Kuwo) Album(id string) (*model.Playlist, []model.Song, error) {
	id = strings.TrimSpace(id)
	if id == "" {
		return nil, nil, fmt.Errorf("kuwo album id is empty")
	}
	endpoint := kuwoSearchBaseURL + "?pn=0&rn=1000&stype=albuminfo&albumid=" + url.QueryEscape(id) +
		"&sortby=0&alflac=1&show_copyright_off=1&pcmp4=1&encoding=utf8"
	payload, err := getSingleQuotedJSON(endpoint, client.Cookie, nil)
	if err != nil {
		return nil, nil, err
	}
	album := kuwoAlbum(payload)
	if album.ID == "" {
		album.ID = id
		album.Link = kuwoAlbumLink(id)
	}
	songs := kuwoSongs(payload["musiclist"], album)
	if len(songs) == 0 {
		return nil, nil, fmt.Errorf("kuwo album response has no songs: musiclist=%T songnum=%s", payload["musiclist"], text(payload["songnum"]))
	}
	if album.TrackCount == 0 {
		album.TrackCount = len(songs)
	}
	return &album, songs, nil
}

func (client *Kuwo) ParsePlaylist(link string) (*model.Playlist, []model.Song, error) {
	matches := kuwoPlaylistLinkPattern.FindStringSubmatch(strings.TrimSpace(link))
	if len(matches) != 2 {
		return nil, nil, fmt.Errorf("invalid kuwo playlist link")
	}
	return client.Playlist(matches[1])
}

func (client *Kuwo) ParseAlbum(link string) (*model.Playlist, []model.Song, error) {
	matches := kuwoAlbumLinkPattern.FindStringSubmatch(strings.TrimSpace(link))
	if len(matches) != 2 {
		return nil, nil, fmt.Errorf("invalid kuwo album link")
	}
	return client.Album(matches[1])
}

func (client *Kuwo) RecommendedPlaylists() ([]model.Playlist, error) {
	query := url.Values{
		"pn": {"1"}, "rn": {"30"}, "order": {"hot"},
		"loginUid": {"0"}, "loginSid": {"0"}, "appUid": {"38668888"},
	}
	payload, err := getJSON(kuwoBrowseBaseURL+"/getRcmPlayList?"+query.Encode(), client.Cookie, nil)
	if err != nil {
		return nil, err
	}
	if integer(payload["code"]) != 200 {
		return nil, fmt.Errorf("kuwo recommendations failed: code=%d message=%s", integer(payload["code"]), text(payload["msg"]))
	}
	playlists := kuwoBrowsePlaylists(at(payload, "data", "data"))
	if len(playlists) == 0 {
		return nil, errorsForEmpty("kuwo recommendations")
	}
	return playlists, nil
}

func (client *Kuwo) PlaylistCategories() ([]model.PlaylistCategory, error) {
	query := url.Values{"loginUid": {"0"}, "loginSid": {"0"}, "appUid": {"38668888"}}
	payload, err := getJSON(kuwoBrowseBaseURL+"/getTagList?"+query.Encode(), client.Cookie, nil)
	if err != nil {
		return nil, err
	}
	if integer(payload["code"]) != 200 {
		return nil, fmt.Errorf("kuwo playlist categories failed: code=%d message=%s", integer(payload["code"]), text(payload["msg"]))
	}
	result := []model.PlaylistCategory{{ID: "", Name: "全部", Group: "全部", Source: "kuwo"}}
	for _, groupValue := range array(payload["data"]) {
		group := object(groupValue)
		groupName := normalizeKuwoText(text(group["name"]))
		for _, itemValue := range array(group["data"]) {
			item := object(itemValue)
			id := text(item["id"])
			name := normalizeKuwoText(text(item["name"]))
			if id == "" || name == "" {
				continue
			}
			result = append(result, model.PlaylistCategory{
				ID: id, Name: name, Group: groupName, Source: "kuwo",
				Hot: strings.Contains(strings.ToUpper(text(item["extend"])), "HOT"),
			})
		}
	}
	return result, nil
}

func (client *Kuwo) CategoryPlaylists(categoryID string, page, limit int) ([]model.Playlist, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 30
	}
	query := url.Values{
		"pn": {strconv.Itoa(page)}, "rn": {strconv.Itoa(limit)},
		"loginUid": {"0"}, "loginSid": {"0"}, "appUid": {"38668888"},
	}
	endpoint := "/getRcmPlayList"
	if strings.TrimSpace(categoryID) == "" {
		query.Set("order", "hot")
	} else {
		endpoint = "/getTagPlayList"
		query.Set("id", strings.TrimSpace(categoryID))
	}
	payload, err := getJSON(kuwoBrowseBaseURL+endpoint+"?"+query.Encode(), client.Cookie, nil)
	if err != nil {
		return nil, err
	}
	if integer(payload["code"]) != 200 {
		return nil, fmt.Errorf("kuwo category playlists failed: code=%d message=%s", integer(payload["code"]), text(payload["msg"]))
	}
	playlists := kuwoBrowsePlaylists(at(payload, "data", "data"))
	if len(playlists) == 0 {
		return nil, errorsForEmpty("kuwo category playlists")
	}
	return playlists, nil
}

func kuwoSongs(value interface{}, collection model.Playlist) []model.Song {
	items := array(value)
	result := make([]model.Song, 0, len(items))
	for _, item := range items {
		raw := object(item)
		id := firstText(raw, "id", "musicrid", "audio_id")
		id = strings.TrimPrefix(id, "MUSIC_")
		if id == "" {
			continue
		}
		albumID := firstText(raw, "albumid", "albumId")
		if albumID == "" {
			albumID = collection.ID
		}
		album := normalizeKuwoText(firstText(raw, "album", "falbum"))
		if album == "" {
			album = collection.Name
		}
		cover := normalizeKuwoImageURL(firstText(raw, "albumpic", "pic120", "web_albumpic_short", "img"))
		if cover == "" {
			cover = collection.Cover
		}
		name := normalizeKuwoText(firstText(raw, "name", "songname", "fsongname"))
		artist := normalizeKuwoText(firstText(raw, "artist", "aartist", "fartist"))
		result = append(result, model.Song{
			ID: id, Name: name, Artist: artist, Album: album, AlbumID: albumID,
			Duration: integer(raw["duration"]), Source: "kuwo", Cover: cover,
			Link: "http://www.kuwo.cn/play_detail/" + id,
			Extra: map[string]string{
				"rid": id, "album_id": albumID, "provider_lookup": strings.TrimSpace(name + " " + artist),
			},
		})
	}
	return result
}

func kuwoPlaylists(value interface{}) []model.Playlist {
	items := array(value)
	result := make([]model.Playlist, 0, len(items))
	for _, item := range items {
		raw := object(item)
		id := text(raw["playlistid"])
		if id == "" {
			continue
		}
		result = append(result, model.Playlist{
			ID: id, Name: normalizeKuwoText(text(raw["name"])), Cover: normalizeKuwoImageURL(text(raw["pic"])),
			TrackCount: integer(raw["songnum"]), PlayCount: integer(raw["playcnt"]),
			Creator: normalizeKuwoText(text(raw["nickname"])), Description: normalizeKuwoText(text(raw["intro"])),
			Source: "kuwo", Link: kuwoPlaylistLink(id),
		})
	}
	return result
}

func kuwoPlaylistDetail(raw map[string]interface{}, fallbackID string) model.Playlist {
	id := firstText(raw, "id", "pid")
	if id == "" {
		id = fallbackID
	}
	return model.Playlist{
		ID: id, Name: normalizeKuwoText(firstText(raw, "title", "name")), Cover: normalizeKuwoImageURL(text(raw["pic"])),
		TrackCount: firstInteger(raw, "total", "validtotal"), PlayCount: integer(raw["playnum"]),
		Creator: normalizeKuwoText(text(raw["uname"])), Description: normalizeKuwoText(text(raw["info"])),
		Source: "kuwo", Link: kuwoPlaylistLink(id),
	}
}

func kuwoAlbums(value interface{}) []model.Playlist {
	items := array(value)
	result := make([]model.Playlist, 0, len(items))
	for _, item := range items {
		album := kuwoAlbum(object(item))
		if album.ID != "" {
			result = append(result, album)
		}
	}
	return result
}

func kuwoAlbum(raw map[string]interface{}) model.Playlist {
	id := firstText(raw, "albumid", "id")
	return model.Playlist{
		ID: id, Name: normalizeKuwoText(firstText(raw, "name", "album")),
		Cover:      normalizeKuwoImageURL(firstText(raw, "hts_img", "img", "pic")),
		TrackCount: firstInteger(raw, "songnum", "musiccnt"), PlayCount: firstInteger(raw, "PLAYCNT", "playcnt"),
		Creator: normalizeKuwoText(firstText(raw, "aartist", "artist")), Description: normalizeKuwoText(text(raw["info"])),
		Source: "kuwo", Link: kuwoAlbumLink(id),
	}
}

func kuwoBrowsePlaylists(value interface{}) []model.Playlist {
	items := array(value)
	result := make([]model.Playlist, 0, len(items))
	for _, item := range items {
		raw := object(item)
		id := text(raw["id"])
		if id == "" {
			continue
		}
		result = append(result, model.Playlist{
			ID: id, Name: normalizeKuwoText(text(raw["name"])), Cover: normalizeKuwoImageURL(text(raw["img"])),
			TrackCount: firstInteger(raw, "songnum", "total", "count", "musicnum"), PlayCount: integer(raw["listencnt"]),
			Creator: normalizeKuwoText(text(raw["uname"])), Description: normalizeKuwoText(firstText(raw, "desc", "info")),
			Source: "kuwo", Link: kuwoPlaylistLink(id),
		})
	}
	return result
}

func normalizeKuwoText(value string) string {
	return strings.TrimSpace(strings.NewReplacer("&nbsp;", " ", "&amp;", "&", "&quot;", "\"").Replace(value))
}

func normalizeKuwoImageURL(value string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return ""
	}
	value = strings.Replace(value, "_150.", "_700.", 1)
	value = strings.Replace(value, "_120.", "_700.", 1)
	value = strings.Replace(value, "_100.", "_700.", 1)
	if strings.HasPrefix(value, "//") {
		return "http:" + value
	}
	if !strings.HasPrefix(value, "http://") && !strings.HasPrefix(value, "https://") {
		return "http://" + strings.TrimPrefix(value, "/")
	}
	return value
}

func kuwoPlaylistLink(id string) string {
	if id == "" {
		return ""
	}
	return "http://www.kuwo.cn/playlist_detail/" + id
}

func kuwoAlbumLink(id string) string {
	if id == "" {
		return ""
	}
	return "http://www.kuwo.cn/album_detail/" + id
}

File: backend/provider_bridge/platforms/kuwo.py
"""Kuwo collection adapter hosted by the provider sidecar."""

from __future__ import annotations

import re
from urllib.parse import urlencode

from provider_bridge.collection_common import (
    at,
    category_payload,
    collection_payload,
    first,
    first_integer,
    integer,
    link_id,
    list_value,
    normalize_kuwo_text,
    object_value,
    page_limit,
    string,
    track_payload,
)
from provider_bridge.platform_http import PlatformHTTP


SEARCH_URL = "http://search.kuwo.cn/r.s"
LIST_URL = "http://nplserver.kuwo.cn/pl.svc"
BROWSE_URL = "http://wapi.kuwo.cn/api/pc/classify/playlist"


class KuwoCollections:
    source = "kuwo"

    def __init__(self, cookie: str = "", session=None):
        self.http = PlatformHTTP(cookie, session)

    def search_playlist(self, keyword: str, **_):
        return self._playlists(self._search(keyword, "playlist").get("abslist"))

    def search_album(self, keyword: str, **_):
        return self._albums(self._search(keyword, "album").get("albumlist"))

    def _search(self, keyword: str, collection_type: str):
        query = {
            "client": "kt", "all": string(keyword), "pn": 0, "rn": 30, "uid": 0,
            "ver": "kwplayer_ar_9.2.2.1", "vipver": 1, "show_copyright_off": 1,
            "newver": 1, "ft": collection_type, "cluster": 0, "strategy": 2012,
            "encoding": "utf8", "rformat": "json", "mobi": 1,
        }
        return self.http.get(SEARCH_URL + "?" + urlencode(query))

    def playlist(self, identifier: str, **_):
        identifier = string(identifier)
        if not identifier:
            raise ValueError("kuwo playlist id is empty")
        query = {
            "op": "getlistinfo", "pid": identifier, "pn": 0, "rn": 1000,
            "encode": "utf8", "keyset": "pl2012", "vipver": "MUSIC_9.1.1.2_BCS2", "newver": 1,
        }
        payload = self.http.get(LIST_URL + "?" + urlencode(query))
        playlist = self._playlist_detail(payload, identifier)
        songs = self._songs(payload.get("musiclist"), playlist)
        if not songs:
            raise ValueError("kuwo playlist response has no songs")
        if not playlist["track_count"]:
            playlist["track_count"] = len(songs)
        return playlist, songs

    def album(self, identifier: str, **_):
        identifier = string(identifier)
        if not identifier:
            raise ValueError("kuwo album id is empty")
        query = {
            "pn": 0, "rn": 1000, "stype": "albuminfo", "albumid": identifier,
            "sortby": 0, "alflac": 1, "show_copyright_off": 1, "pcmp4": 1, "encoding": "utf8",
        }
        payload = self.http.get(SEARCH_URL + "?" + urlencode(query))
        album = self._album(object_value(payload))
        album["id"] = album["id"] or identifier
        album["link"] = self._album_link(album["id"])
        songs = self._songs(payload.get("musiclist"), album)
        if not songs:
            raise ValueError("kuwo album response has no songs")
        if not album["track_count"]:
            album["track_count"] = len(songs)
        return album, songs

    def parse_playlist(self, link: str, **_):
        identifier = link_id(link)
        if not re.search(r"/playlist_detail/\d+", string(link), re.IGNORECASE) or not identifier:
            raise ValueError("invalid kuwo playlist link")
        return self.playlist(identifier)

    def parse_album(self, link: str, **_):
        identifier = link_id(link)
        if not re.search(r"/album_detail/\d+", string(link), re.IGNORECASE) or not identifier:
            raise ValueError("invalid kuwo album link")
        return self.album(identifier)

    def recommend(self, **_):
        return self.category("", 1, 30)

    def categories(self, **_):
        payload = self.http.get(BROWSE_URL + "/getTagList?" + urlencode({"loginUid": 0, "loginSid": 0, "appUid": 38668888}))
        if integer(payload.get("code")) != 200:
            raise ValueError(f"kuwo playlist categories failed: code={integer(payload.get('code'))}")
        result = [category_payload(identifier="", source=self.source, name="全部", group="全部")]
        for group_value in list_value(payload.get("data")):
            group = object_value(group_value)
            group_name = normalize_kuwo_text(group.get("name"))
            for item_value in list_value(group.get("data")):
                item = object_value(item_value)
                identifier = string(item.get("id"))
                name = normalize_kuwo_text(item.get("name"))
                if identifier and name:
                    result.append(category_payload(
                        identifier=identifier, source=self.source, name=name, group=group_name,
                        hot="HOT" in string(item.get("extend")).upper(),
                    ))
        return result

    def category(self, category_id: str, page: int = 1, limit: int = 30, **_):
        page, limit = page_limit(page, limit)
        query = {"pn": page, "rn": limit, "loginUid": 0, "loginSid": 0, "appUid": 38668888}
        endpoint = "/getRcmPlayList"
        if string(category_id):
            endpoint = "/getTagPlayList"
            query["id"] = string(category_id)
        else:
            query["order"] = "hot"
        payload = self.http.get(BROWSE_URL + endpoint + "?" + urlencode(query))
        if integer(payload.get("code")) != 200:
            raise ValueError(f"kuwo category playlists failed: code={integer(payload.get('code'))}")
        playlists = self._browse_playlists(at(payload, "data", "data"))
        if not playlists:
            raise ValueError("kuwo category playlists response is empty")
        return playlists

    def _songs(self, value, collection):
        result = []
        for raw_value in list_value(value):
            raw = object_value(raw_value)
            identifier = first(raw, "id", "musicrid", "audio_id").removeprefix("MUSIC_")
            if not identifier:
                continue
            album_id = first(raw, "albumid", "albumId") or collection["id"]
            album = normalize_kuwo_text(first(raw, "album", "falbum")) or collection["name"]
            cover = self._image(first(raw, "albumpic", "pic120", "web_albumpic_short", "img")) or collection["cover"]
            name = normalize_kuwo_text(first(raw, "name", "songname", "fsongname"))
            artist = normalize_kuwo_text(first(raw, "artist", "aartist", "fartist"))
            result.append(track_payload(
                identifier=identifier, source=self.source, name=name, artist=artist,
                album=album, album_id=album_id, duration=integer(raw.get("duration")), cover=cover,
                link="http://www.kuwo.cn/play_detail/" + identifier,
                extra={"rid": identifier, "album_id": album_id, "provider_lookup": f"{name} {artist}".strip()},
            ))
        return result

    def _playlists(self, value):
        result = []
        for raw_value in list_value(value):
            raw = object_value(raw_value)
            identifier = string(raw.get("playlistid"))
            if identifier:
                result.append(collection_payload(
                    identifier=identifier, source=self.source, name=normalize_kuwo_text(raw.get("name")),
                    cover=self._image(raw.get("pic")), track_count=integer(raw.get("songnum")),
                    play_count=integer(raw.get("playcnt")), creator=normalize_kuwo_text(raw.get("nickname")),
                    description=normalize_kuwo_text(raw.get("intro")), link=self._playlist_link(identifier),
                ))
        return result

    def _browse_playlists(self, value):
        result = []
        for raw_value in list_value(value):
            raw = object_value(raw_value)
            identifier = string(raw.get("id"))
            if identifier:
                result.append(collection_payload(
                    identifier=identifier, source=self.source, name=normalize_kuwo_text(raw.get("name")),
                    cover=self._image(raw.get("img")), track_count=first_integer(raw, "songnum", "total", "count", "musicnum"),
                    play_count=integer(raw.get("listencnt")), creator=normalize_kuwo_text(raw.get("uname")),
                    description=normalize_kuwo_text(first(raw, "desc", "info")), link=self._playlist_link(identifier),
                ))
        return result

    def _playlist_detail(self, raw, fallback_id):
        identifier = first(raw, "id", "pid") or fallback_id
        return collection_payload(
            identifier=identifier, source=self.source, name=normalize_kuwo_text(first(raw, "title", "name")),
            cover=self._image(raw.get("pic")), track_count=first_integer(raw, "total", "validtotal"),
            play_count=integer(raw.get("playnum")), creator=normalize_kuwo_text(raw.get("uname")),
            description=normalize_kuwo_text(raw.get("info")), link=self._playlist_link(identifier),
        )

    def _albums(self, value):
        result = []
        for raw_value in list_value(value):
            raw = object_value(raw_value)
            identifier = first(raw, "albumid", "id")
            if identifier:
                result.append(self._album(raw))
        return result

    def _album(self, raw):
        identifier = first(raw, "albumid", "id")
        return collection_payload(
            identifier=identifier, source=self.source, name=normalize_kuwo_text(first(raw, "name", "album")),
            cover=self._image(first(raw, "hts_img", "img", "pic")),
            track_count=first_integer(raw, "songnum", "musiccnt"), play_count=first_integer(raw, "PLAYCNT", "playcnt"),
            creator=normalize_kuwo_text(first(raw, "aartist", "artist")), description=normalize_kuwo_text(raw.get("info")),
            link=self._album_link(identifier),
        )

    @staticmethod
    def _image(value):
        value = string(value)
        for old in ("_150.", "_120.", "_100."):
            value = value.replace(old, "_700.", 1)
        if value.startswith("//"):
            return "http:" + value
        if value and not value.startswith(("http://", "https://")):
            return "http://" + value.lstrip("/")
        return value

    @staticmethod
    def _playlist_link(identifier):
        return "http://www.kuwo.cn/playlist_detail/" + identifier if identifier else ""

    @staticmethod
    def _album_link(identifier):
        return "http://www.kuwo.cn/album_detail/" + identifier if identifier else ""

